google/gax/__init__.py

```diff
@@ -482,3 +491,157 @@ def __next__(self):
+                that are used when reloading the operation.
+        """
+        self._last_operation = operation
+        self._client = client
+        self._result_type = result_type
+        self._metadata_type = metadata_type
+        self._call_options = call_options
+        self._done_clbks = []
+        self._process = None
+
+    def cancel(self):
+        """If last Operation's value of `done` is true, returns false;
+           otherwise, issues OperationsClient.cancel_operation and returns true.
+        """
+        if self.done():
+            return False
+
+        self._client.cancel_operation(self._last_operation.name)
+        return True
+
+    def result(self, timeout=None):
+        """Enters polling loop on OperationsClient.get_operation, and once
+           Operation.done is true, then returns Operation.response if successful
+           or throws GaxError if not successful.
+
+           This method will wait up to timeout seconds. If the call hasn't
+           completed in timeout seconds, then a concurrent.futures.TimeoutError
+           will be raised. timeout can be an int or float. If timeout is not
+           specified or None, there is no limit to the wait time.
+        """
+        if not self._poll(timeout).HasField('response'):
+            raise GaxError(self._last_operation.error.message)
+
+        return _from_any(self._result_type, self._last_operation.response)
+
+    def exception(self, timeout=None):
+        """Similar to result(), except returns the exception if any."""
+        if self._poll(timeout).HasField('error'):
+            return self._last_operation.error
+
+        return None
+
+    def cancelled(self):
+        """Return True if the call was successfully cancelled."""
+        self._get_operation()
+        return self._last_operation.HasField('error') and \
+            self._last_operation.error.code == code_pb2.CANCELLED
+
+    def done(self):
+        """Issues OperationsClient.get_operation and returns value of
+           Operation.done.
+        """
+        return self._get_operation().done
+
+    def add_done_callback(self, fn):  # pylint: disable=invalid-name
+        """Enters a polling loop on OperationsClient.get_operation, and once the
+           operation is done or cancelled, calls the function with this
+           _OperationFuture.
+        """
+        def _execute_clbks(self):
+            self._poll()
+            for clbk in self._done_clbks:
+                try:
+                    clbk(self)
+                except Exception as ex:  # pylint: disable=broad-except
+                    _LOG.exception(ex)
+
+        self._done_clbks.append(fn)
+        if self._process is None:
+            self._process = multiprocessing.Process(
+                target=_execute_clbks, args=(self,))
+            self._process.start()
+        elif not self._process.is_alive() and self._last_operation.done:
+            _execute_clbks(self)
```

Comment:
Be careful here. This can cause some callbacks to be called twice since execute_clbks always calls every callback in the list.

For example:
Register cb A.
Register cb B.
Polling Completes.
Run cb A and cb B.
Some time.
Register cb C.
Run cb A, cb B, cb C